{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# iModulonDB Tutorial\n",
    "\n",
    "This notebook will help you add information and generate \n",
    "\n",
    "Follow the advice [here](https://github.com/SBRG/modulytics/wiki/Adding-a-Private-Project) to get started, and learn to use pymodulon with the initial Tutorial if needed.\n",
    "\n",
    "Please contact krychel@eng.ucsd.edu with any questions!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "\n",
    "from pymodulon.core import IcaData\n",
    "from pymodulon.io import *\n",
    "from pymodulon.imodulondb import *\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 1: Load Your Data\n",
    "\n",
    "More details on this can be found in the main tutorial.\n",
    "\n",
    "All that is absolutely required is the following:\n",
    "\n",
    "- M matrix (part of initial ICA output)\n",
    "- A matrix (part of initial ICA output)\n",
    "- X matrix (log_tpm.csv) -- initial data, preferably not normalized to the base condition.\n",
    "- Sample_table containing:\n",
    "    - Either a column labeled 'sample' or an index (leftmost column) that serves the purpose of sample labels. These should be unique to each sample (no duplicates in the entire table), and it must match the columns of A and X.\n",
    "    - A column labeled 'project' -- this is the largest grouping of samples; they may be downloaded from the same original paper or contain a set of samples/experiments with a common theme. The contents of this column will appear on the X axis when zoomed all the way out in the activity bar graphs on iModulonDB.\n",
    "    - A column labeled 'condition' -- this groups samples such that all biological replicates have the same condition name. Samples in the same project with matching conditions will be grouped into a single bar in the activity bar graphs. If you'd like to re-use condition names across different projects (for example, giving each project a 'control' condition), that is okay.\n",
    "    \n",
    "Additional information will only improve your version of the site!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "ica_data = load_json_model(\"../../mtb_ica.json\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TF</th>\n",
       "      <th>Function</th>\n",
       "      <th>Category</th>\n",
       "      <th>n_genes</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>category_num</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SigH</th>\n",
       "      <td>sigH</td>\n",
       "      <td>Sigma Factor involved in heat shock and oxidat...</td>\n",
       "      <td>Virulence/Persistence</td>\n",
       "      <td>45</td>\n",
       "      <td>0.082192</td>\n",
       "      <td>0.800000</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Polyketide Synthase Complex</th>\n",
       "      <td>None</td>\n",
       "      <td>KEGG Map related to the synthesis of type II p...</td>\n",
       "      <td>Fatty Acid/Cholestrol</td>\n",
       "      <td>2</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IniR</th>\n",
       "      <td>iniR</td>\n",
       "      <td>Transcription factor that regulates the expres...</td>\n",
       "      <td>Antibiotic</td>\n",
       "      <td>3</td>\n",
       "      <td>0.100000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PyrR</th>\n",
       "      <td>pyrR</td>\n",
       "      <td>Transcription factor that regulates the expres...</td>\n",
       "      <td>Nucleic Acids</td>\n",
       "      <td>44</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WhiB4</th>\n",
       "      <td>whiB4</td>\n",
       "      <td>Redox-Responsive transcription regulator that ...</td>\n",
       "      <td>Redox</td>\n",
       "      <td>95</td>\n",
       "      <td>0.184834</td>\n",
       "      <td>0.410526</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Unc_9</th>\n",
       "      <td>None</td>\n",
       "      <td>Uncharacterized iModulon 9</td>\n",
       "      <td>Uncharacterized</td>\n",
       "      <td>24</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SG_13</th>\n",
       "      <td>None</td>\n",
       "      <td>Single Gene iModulon 13</td>\n",
       "      <td>Single Gene</td>\n",
       "      <td>5</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>76</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Central Carbon Metabolism</th>\n",
       "      <td>None</td>\n",
       "      <td>iModulon enriched with the Central Carbon Meta...</td>\n",
       "      <td>Central Carbon</td>\n",
       "      <td>28</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>77</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Positive Regulation of Growth</th>\n",
       "      <td>None</td>\n",
       "      <td>iModulon enriched with the Positive Regulation...</td>\n",
       "      <td>Growth</td>\n",
       "      <td>26</td>\n",
       "      <td>0.366667</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>78</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WhiB4/IdeR</th>\n",
       "      <td>whiB4/ideR</td>\n",
       "      <td>iModulon that maps to the union of the Rv3681c...</td>\n",
       "      <td>Redox</td>\n",
       "      <td>16</td>\n",
       "      <td>0.230769</td>\n",
       "      <td>0.375000</td>\n",
       "      <td>79</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>80 rows Ã— 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                       TF  \\\n",
       "SigH                                 sigH   \n",
       "Polyketide Synthase Complex          None   \n",
       "IniR                                 iniR   \n",
       "PyrR                                 pyrR   \n",
       "WhiB4                               whiB4   \n",
       "...                                   ...   \n",
       "Unc_9                                None   \n",
       "SG_13                                None   \n",
       "Central Carbon Metabolism            None   \n",
       "Positive Regulation of Growth        None   \n",
       "WhiB4/IdeR                     whiB4/ideR   \n",
       "\n",
       "                                                                        Function  \\\n",
       "SigH                           Sigma Factor involved in heat shock and oxidat...   \n",
       "Polyketide Synthase Complex    KEGG Map related to the synthesis of type II p...   \n",
       "IniR                           Transcription factor that regulates the expres...   \n",
       "PyrR                           Transcription factor that regulates the expres...   \n",
       "WhiB4                          Redox-Responsive transcription regulator that ...   \n",
       "...                                                                          ...   \n",
       "Unc_9                                                 Uncharacterized iModulon 9   \n",
       "SG_13                                                    Single Gene iModulon 13   \n",
       "Central Carbon Metabolism      iModulon enriched with the Central Carbon Meta...   \n",
       "Positive Regulation of Growth  iModulon enriched with the Positive Regulation...   \n",
       "WhiB4/IdeR                     iModulon that maps to the union of the Rv3681c...   \n",
       "\n",
       "                                            Category  n_genes  precision  \\\n",
       "SigH                           Virulence/Persistence       45   0.082192   \n",
       "Polyketide Synthase Complex    Fatty Acid/Cholestrol        2   0.333333   \n",
       "IniR                                      Antibiotic        3   0.100000   \n",
       "PyrR                                   Nucleic Acids       44   0.000000   \n",
       "WhiB4                                          Redox       95   0.184834   \n",
       "...                                              ...      ...        ...   \n",
       "Unc_9                                Uncharacterized       24   0.000000   \n",
       "SG_13                                    Single Gene        5   0.000000   \n",
       "Central Carbon Metabolism             Central Carbon       28   0.000000   \n",
       "Positive Regulation of Growth                 Growth       26   0.366667   \n",
       "WhiB4/IdeR                                     Redox       16   0.230769   \n",
       "\n",
       "                                 recall  category_num  \n",
       "SigH                           0.800000             0  \n",
       "Polyketide Synthase Complex    1.000000             1  \n",
       "IniR                           1.000000             2  \n",
       "PyrR                           0.000000             3  \n",
       "WhiB4                          0.410526             4  \n",
       "...                                 ...           ...  \n",
       "Unc_9                          0.000000            75  \n",
       "SG_13                          0.000000            76  \n",
       "Central Carbon Metabolism      0.000000            77  \n",
       "Positive Regulation of Growth  0.333333            78  \n",
       "WhiB4/IdeR                     0.375000            79  \n",
       "\n",
       "[80 rows x 7 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ica_data.imodulon_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>regulator</th>\n",
       "      <th>regulator_id</th>\n",
       "      <th>evidence</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>fhaB</td>\n",
       "      <td>Rv0019c</td>\n",
       "      <td>ChipSeq</td>\n",
       "      <td>Rv1733c</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>fhaA</td>\n",
       "      <td>Rv0020c</td>\n",
       "      <td>ChipSeq</td>\n",
       "      <td>Rv0020c</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>fhaA</td>\n",
       "      <td>Rv0020c</td>\n",
       "      <td>ChipSeq</td>\n",
       "      <td>Rv1733c</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>whiB5</td>\n",
       "      <td>Rv0022c</td>\n",
       "      <td>ChipSeq</td>\n",
       "      <td>Rv0013</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>whiB5</td>\n",
       "      <td>Rv0022c</td>\n",
       "      <td>ChipSeq</td>\n",
       "      <td>Rv0038</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24300</th>\n",
       "      <td>24300</td>\n",
       "      <td>lysG</td>\n",
       "      <td>Rv1985c</td>\n",
       "      <td>Differential Expression</td>\n",
       "      <td>Rv2932</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24301</th>\n",
       "      <td>24301</td>\n",
       "      <td>lysG</td>\n",
       "      <td>Rv1985c</td>\n",
       "      <td>Differential Expression</td>\n",
       "      <td>Rv2933</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24302</th>\n",
       "      <td>24302</td>\n",
       "      <td>lysG</td>\n",
       "      <td>Rv1985c</td>\n",
       "      <td>Differential Expression</td>\n",
       "      <td>Rv2934</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24303</th>\n",
       "      <td>24303</td>\n",
       "      <td>lysG</td>\n",
       "      <td>Rv1985c</td>\n",
       "      <td>Differential Expression</td>\n",
       "      <td>Rv1986</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24304</th>\n",
       "      <td>24304</td>\n",
       "      <td>lysG</td>\n",
       "      <td>Rv1985c</td>\n",
       "      <td>Differential Expression</td>\n",
       "      <td>Rv1985c</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>24305 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Unnamed: 0 regulator regulator_id                 evidence  gene_id\n",
       "0               0      fhaB      Rv0019c                  ChipSeq  Rv1733c\n",
       "1               1      fhaA      Rv0020c                  ChipSeq  Rv0020c\n",
       "2               2      fhaA      Rv0020c                  ChipSeq  Rv1733c\n",
       "3               3     whiB5      Rv0022c                  ChipSeq   Rv0013\n",
       "4               4     whiB5      Rv0022c                  ChipSeq   Rv0038\n",
       "...           ...       ...          ...                      ...      ...\n",
       "24300       24300      lysG      Rv1985c  Differential Expression   Rv2932\n",
       "24301       24301      lysG      Rv1985c  Differential Expression   Rv2933\n",
       "24302       24302      lysG      Rv1985c  Differential Expression   Rv2934\n",
       "24303       24303      lysG      Rv1985c  Differential Expression   Rv1986\n",
       "24304       24304      lysG      Rv1985c  Differential Expression  Rv1985c\n",
       "\n",
       "[24305 rows x 5 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ica_data.trn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SigH</th>\n",
       "      <th>Polyketide Synthase Complex</th>\n",
       "      <th>IniR</th>\n",
       "      <th>PyrR</th>\n",
       "      <th>WhiB4</th>\n",
       "      <th>DevR-1</th>\n",
       "      <th>MarR</th>\n",
       "      <th>GroEL-GroES Complex</th>\n",
       "      <th>IdeR</th>\n",
       "      <th>SigK</th>\n",
       "      <th>...</th>\n",
       "      <th>Lsr2</th>\n",
       "      <th>PDIM;PGL Synthesis</th>\n",
       "      <th>Unc_8</th>\n",
       "      <th>SG_11</th>\n",
       "      <th>SG_12</th>\n",
       "      <th>Unc_9</th>\n",
       "      <th>SG_13</th>\n",
       "      <th>Central Carbon Metabolism</th>\n",
       "      <th>Positive Regulation of Growth</th>\n",
       "      <th>WhiB4/IdeR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Rv0001</th>\n",
       "      <td>-0.002362</td>\n",
       "      <td>-0.000345</td>\n",
       "      <td>0.018863</td>\n",
       "      <td>0.003751</td>\n",
       "      <td>-0.004513</td>\n",
       "      <td>-0.005628</td>\n",
       "      <td>-0.008293</td>\n",
       "      <td>-0.001941</td>\n",
       "      <td>0.004094</td>\n",
       "      <td>-0.003420</td>\n",
       "      <td>...</td>\n",
       "      <td>0.043023</td>\n",
       "      <td>0.021328</td>\n",
       "      <td>-0.001038</td>\n",
       "      <td>-0.006600</td>\n",
       "      <td>0.014561</td>\n",
       "      <td>0.000662</td>\n",
       "      <td>-0.001623</td>\n",
       "      <td>-0.023295</td>\n",
       "      <td>-0.006675</td>\n",
       "      <td>-0.006532</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rv0002</th>\n",
       "      <td>-0.005701</td>\n",
       "      <td>0.003088</td>\n",
       "      <td>-0.002460</td>\n",
       "      <td>0.004126</td>\n",
       "      <td>0.004209</td>\n",
       "      <td>-0.004811</td>\n",
       "      <td>-0.014755</td>\n",
       "      <td>-0.009028</td>\n",
       "      <td>0.002528</td>\n",
       "      <td>-0.004720</td>\n",
       "      <td>...</td>\n",
       "      <td>0.020308</td>\n",
       "      <td>0.005312</td>\n",
       "      <td>-0.020528</td>\n",
       "      <td>-0.007400</td>\n",
       "      <td>0.003026</td>\n",
       "      <td>-0.005617</td>\n",
       "      <td>0.000873</td>\n",
       "      <td>-0.005503</td>\n",
       "      <td>-0.006236</td>\n",
       "      <td>-0.010049</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rv0003</th>\n",
       "      <td>-0.015815</td>\n",
       "      <td>0.002336</td>\n",
       "      <td>0.009974</td>\n",
       "      <td>0.001325</td>\n",
       "      <td>0.005892</td>\n",
       "      <td>-0.010023</td>\n",
       "      <td>-0.003943</td>\n",
       "      <td>-0.013108</td>\n",
       "      <td>0.000750</td>\n",
       "      <td>-0.001532</td>\n",
       "      <td>...</td>\n",
       "      <td>0.026046</td>\n",
       "      <td>-0.000131</td>\n",
       "      <td>-0.005501</td>\n",
       "      <td>0.003283</td>\n",
       "      <td>0.002269</td>\n",
       "      <td>-0.001001</td>\n",
       "      <td>-0.003914</td>\n",
       "      <td>-0.004735</td>\n",
       "      <td>-0.003045</td>\n",
       "      <td>-0.011112</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rv0004</th>\n",
       "      <td>-0.010953</td>\n",
       "      <td>0.014130</td>\n",
       "      <td>0.005316</td>\n",
       "      <td>0.007447</td>\n",
       "      <td>0.002355</td>\n",
       "      <td>-0.005675</td>\n",
       "      <td>0.001232</td>\n",
       "      <td>0.002642</td>\n",
       "      <td>-0.002937</td>\n",
       "      <td>-0.009452</td>\n",
       "      <td>...</td>\n",
       "      <td>0.002391</td>\n",
       "      <td>-0.000943</td>\n",
       "      <td>0.047164</td>\n",
       "      <td>0.000554</td>\n",
       "      <td>-0.010171</td>\n",
       "      <td>-0.001388</td>\n",
       "      <td>-0.008658</td>\n",
       "      <td>-0.018899</td>\n",
       "      <td>0.004760</td>\n",
       "      <td>-0.013165</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rv0005</th>\n",
       "      <td>-0.009394</td>\n",
       "      <td>-0.002810</td>\n",
       "      <td>-0.012055</td>\n",
       "      <td>0.020868</td>\n",
       "      <td>-0.005547</td>\n",
       "      <td>-0.001030</td>\n",
       "      <td>-0.001101</td>\n",
       "      <td>0.003174</td>\n",
       "      <td>-0.003904</td>\n",
       "      <td>0.000795</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.000266</td>\n",
       "      <td>-0.003107</td>\n",
       "      <td>0.000432</td>\n",
       "      <td>-0.003614</td>\n",
       "      <td>0.001742</td>\n",
       "      <td>-0.003190</td>\n",
       "      <td>0.006061</td>\n",
       "      <td>-0.010431</td>\n",
       "      <td>-0.001558</td>\n",
       "      <td>-0.010690</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rv3920c</th>\n",
       "      <td>-0.010412</td>\n",
       "      <td>0.013396</td>\n",
       "      <td>-0.011807</td>\n",
       "      <td>0.000782</td>\n",
       "      <td>-0.011104</td>\n",
       "      <td>-0.004583</td>\n",
       "      <td>-0.008442</td>\n",
       "      <td>-0.010533</td>\n",
       "      <td>-0.000110</td>\n",
       "      <td>-0.012717</td>\n",
       "      <td>...</td>\n",
       "      <td>0.004892</td>\n",
       "      <td>-0.008485</td>\n",
       "      <td>-0.000913</td>\n",
       "      <td>-0.001425</td>\n",
       "      <td>-0.002206</td>\n",
       "      <td>-0.002644</td>\n",
       "      <td>-0.007711</td>\n",
       "      <td>-0.019890</td>\n",
       "      <td>0.003186</td>\n",
       "      <td>0.000633</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rv3921c</th>\n",
       "      <td>-0.012022</td>\n",
       "      <td>0.005247</td>\n",
       "      <td>-0.011139</td>\n",
       "      <td>0.004236</td>\n",
       "      <td>-0.006812</td>\n",
       "      <td>-0.001795</td>\n",
       "      <td>-0.016140</td>\n",
       "      <td>-0.010954</td>\n",
       "      <td>0.001175</td>\n",
       "      <td>-0.006727</td>\n",
       "      <td>...</td>\n",
       "      <td>0.033010</td>\n",
       "      <td>-0.001548</td>\n",
       "      <td>0.000063</td>\n",
       "      <td>-0.000003</td>\n",
       "      <td>0.005848</td>\n",
       "      <td>0.004321</td>\n",
       "      <td>-0.001381</td>\n",
       "      <td>-0.022975</td>\n",
       "      <td>0.004075</td>\n",
       "      <td>-0.000829</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rv3922c</th>\n",
       "      <td>-0.009557</td>\n",
       "      <td>0.003253</td>\n",
       "      <td>0.005597</td>\n",
       "      <td>0.000352</td>\n",
       "      <td>-0.003766</td>\n",
       "      <td>-0.000583</td>\n",
       "      <td>-0.017275</td>\n",
       "      <td>0.002083</td>\n",
       "      <td>0.002723</td>\n",
       "      <td>0.000216</td>\n",
       "      <td>...</td>\n",
       "      <td>0.024754</td>\n",
       "      <td>0.001314</td>\n",
       "      <td>0.000270</td>\n",
       "      <td>0.014457</td>\n",
       "      <td>0.011590</td>\n",
       "      <td>0.001659</td>\n",
       "      <td>-0.002904</td>\n",
       "      <td>-0.009478</td>\n",
       "      <td>0.014821</td>\n",
       "      <td>-0.003982</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rv3923c</th>\n",
       "      <td>-0.002366</td>\n",
       "      <td>0.004769</td>\n",
       "      <td>-0.005973</td>\n",
       "      <td>0.000878</td>\n",
       "      <td>-0.007411</td>\n",
       "      <td>-0.003503</td>\n",
       "      <td>-0.015604</td>\n",
       "      <td>-0.003700</td>\n",
       "      <td>0.000320</td>\n",
       "      <td>-0.000618</td>\n",
       "      <td>...</td>\n",
       "      <td>0.015781</td>\n",
       "      <td>-0.005540</td>\n",
       "      <td>0.002152</td>\n",
       "      <td>0.003105</td>\n",
       "      <td>0.006952</td>\n",
       "      <td>0.009292</td>\n",
       "      <td>-0.012886</td>\n",
       "      <td>-0.007947</td>\n",
       "      <td>-0.005386</td>\n",
       "      <td>-0.002804</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rv3924c</th>\n",
       "      <td>-0.007361</td>\n",
       "      <td>0.014936</td>\n",
       "      <td>-0.015652</td>\n",
       "      <td>0.001195</td>\n",
       "      <td>0.008647</td>\n",
       "      <td>0.002114</td>\n",
       "      <td>-0.012987</td>\n",
       "      <td>-0.008434</td>\n",
       "      <td>0.001023</td>\n",
       "      <td>-0.007548</td>\n",
       "      <td>...</td>\n",
       "      <td>0.005015</td>\n",
       "      <td>-0.004982</td>\n",
       "      <td>0.004016</td>\n",
       "      <td>0.013488</td>\n",
       "      <td>-0.005106</td>\n",
       "      <td>-0.000890</td>\n",
       "      <td>0.002154</td>\n",
       "      <td>-0.007196</td>\n",
       "      <td>0.000834</td>\n",
       "      <td>0.000804</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3906 rows Ã— 80 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             SigH  Polyketide Synthase Complex      IniR      PyrR     WhiB4  \\\n",
       "Rv0001  -0.002362                    -0.000345  0.018863  0.003751 -0.004513   \n",
       "Rv0002  -0.005701                     0.003088 -0.002460  0.004126  0.004209   \n",
       "Rv0003  -0.015815                     0.002336  0.009974  0.001325  0.005892   \n",
       "Rv0004  -0.010953                     0.014130  0.005316  0.007447  0.002355   \n",
       "Rv0005  -0.009394                    -0.002810 -0.012055  0.020868 -0.005547   \n",
       "...           ...                          ...       ...       ...       ...   \n",
       "Rv3920c -0.010412                     0.013396 -0.011807  0.000782 -0.011104   \n",
       "Rv3921c -0.012022                     0.005247 -0.011139  0.004236 -0.006812   \n",
       "Rv3922c -0.009557                     0.003253  0.005597  0.000352 -0.003766   \n",
       "Rv3923c -0.002366                     0.004769 -0.005973  0.000878 -0.007411   \n",
       "Rv3924c -0.007361                     0.014936 -0.015652  0.001195  0.008647   \n",
       "\n",
       "           DevR-1      MarR  GroEL-GroES Complex      IdeR      SigK  ...  \\\n",
       "Rv0001  -0.005628 -0.008293            -0.001941  0.004094 -0.003420  ...   \n",
       "Rv0002  -0.004811 -0.014755            -0.009028  0.002528 -0.004720  ...   \n",
       "Rv0003  -0.010023 -0.003943            -0.013108  0.000750 -0.001532  ...   \n",
       "Rv0004  -0.005675  0.001232             0.002642 -0.002937 -0.009452  ...   \n",
       "Rv0005  -0.001030 -0.001101             0.003174 -0.003904  0.000795  ...   \n",
       "...           ...       ...                  ...       ...       ...  ...   \n",
       "Rv3920c -0.004583 -0.008442            -0.010533 -0.000110 -0.012717  ...   \n",
       "Rv3921c -0.001795 -0.016140            -0.010954  0.001175 -0.006727  ...   \n",
       "Rv3922c -0.000583 -0.017275             0.002083  0.002723  0.000216  ...   \n",
       "Rv3923c -0.003503 -0.015604            -0.003700  0.000320 -0.000618  ...   \n",
       "Rv3924c  0.002114 -0.012987            -0.008434  0.001023 -0.007548  ...   \n",
       "\n",
       "             Lsr2  PDIM;PGL Synthesis     Unc_8     SG_11     SG_12     Unc_9  \\\n",
       "Rv0001   0.043023            0.021328 -0.001038 -0.006600  0.014561  0.000662   \n",
       "Rv0002   0.020308            0.005312 -0.020528 -0.007400  0.003026 -0.005617   \n",
       "Rv0003   0.026046           -0.000131 -0.005501  0.003283  0.002269 -0.001001   \n",
       "Rv0004   0.002391           -0.000943  0.047164  0.000554 -0.010171 -0.001388   \n",
       "Rv0005  -0.000266           -0.003107  0.000432 -0.003614  0.001742 -0.003190   \n",
       "...           ...                 ...       ...       ...       ...       ...   \n",
       "Rv3920c  0.004892           -0.008485 -0.000913 -0.001425 -0.002206 -0.002644   \n",
       "Rv3921c  0.033010           -0.001548  0.000063 -0.000003  0.005848  0.004321   \n",
       "Rv3922c  0.024754            0.001314  0.000270  0.014457  0.011590  0.001659   \n",
       "Rv3923c  0.015781           -0.005540  0.002152  0.003105  0.006952  0.009292   \n",
       "Rv3924c  0.005015           -0.004982  0.004016  0.013488 -0.005106 -0.000890   \n",
       "\n",
       "            SG_13  Central Carbon Metabolism  Positive Regulation of Growth  \\\n",
       "Rv0001  -0.001623                  -0.023295                      -0.006675   \n",
       "Rv0002   0.000873                  -0.005503                      -0.006236   \n",
       "Rv0003  -0.003914                  -0.004735                      -0.003045   \n",
       "Rv0004  -0.008658                  -0.018899                       0.004760   \n",
       "Rv0005   0.006061                  -0.010431                      -0.001558   \n",
       "...           ...                        ...                            ...   \n",
       "Rv3920c -0.007711                  -0.019890                       0.003186   \n",
       "Rv3921c -0.001381                  -0.022975                       0.004075   \n",
       "Rv3922c -0.002904                  -0.009478                       0.014821   \n",
       "Rv3923c -0.012886                  -0.007947                      -0.005386   \n",
       "Rv3924c  0.002154                  -0.007196                       0.000834   \n",
       "\n",
       "         WhiB4/IdeR  \n",
       "Rv0001    -0.006532  \n",
       "Rv0002    -0.010049  \n",
       "Rv0003    -0.011112  \n",
       "Rv0004    -0.013165  \n",
       "Rv0005    -0.010690  \n",
       "...             ...  \n",
       "Rv3920c    0.000633  \n",
       "Rv3921c   -0.000829  \n",
       "Rv3922c   -0.003982  \n",
       "Rv3923c   -0.002804  \n",
       "Rv3924c    0.000804  \n",
       "\n",
       "[3906 rows x 80 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ica_data.M"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 2: Fill out some metadata\n",
    "\n",
    "When you load an ica_data model, it fills out default metadata for you. Therefore you can technically skip this step.\n",
    "\n",
    "### 2.a: Splash Table (Dictionary)\n",
    "\n",
    "This table provides the basic elements shown on [the splash page](http://imodulondb.org) and the names of the directories in which all data will be stored.\n",
    "\n",
    "First, look at the default one:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'large_title': 'New Dataset',\n",
       " 'subtitle': 'Unpublished study',\n",
       " 'author': 'Pymodulon User',\n",
       " 'organism_folder': 'new_org',\n",
       " 'dataset_folder': 'new_dataset'}"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ica_data.splash_table"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The first three entries are shown on the splash page directly. The last two are important folder names. Typical organism folder names are in the form 'g_species'. Choose a name for your dataset that will be unique; it is likely the only one for your organism, so you could choose 'modulome' if that is the project you are working on.\n",
    "\n",
    "Let's update this table with your information."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'large_title': '<i>M. tuberculosis</i>',\n",
       " 'subtitle': 'M. Tuberculosis Dataset',\n",
       " 'author': 'Yoo, et al.',\n",
       " 'organism_folder': 'm_tuberculosis',\n",
       " 'dataset_folder': 'modulome'}"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# FILL IN THESE DETAILS\n",
    "\n",
    "# <i> tags italicize the species name\n",
    "# So far on the site, large titles are all simply species names.\n",
    "ica_data.splash_table['large_title'] = '<i>M. tuberculosis</i>'\n",
    "\n",
    "# subtitle provides additional information about the dataset's origins\n",
    "ica_data.splash_table['subtitle'] = 'M. Tuberculosis Dataset'\n",
    "ica_data.splash_table['author'] = 'Yoo, et al.'\n",
    "\n",
    "# these will correspond to actual folders, which will be generated for you.\n",
    "ica_data.splash_table['organism_folder'] = 'm_tuberculosis'\n",
    "ica_data.splash_table['dataset_folder'] = 'modulome'\n",
    "\n",
    "# make sure the output is desired\n",
    "ica_data.splash_table"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.b: Dataset Table (Dictionary)\n",
    "\n",
    "This table provides the basic elements shown in the [dataset page](https://imodulondb.org/dataset.html?organism=e_coli&dataset=precise1). The numerical elements are computed for you in the default.\n",
    "\n",
    "If desired, you can add your own entries to this with whatever names you'd like. They can even include links if you obey the [rules for making HTML links](https://www.w3schools.com/html/html_links.asp). For example, the published datasets include a 'Publication' entry."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Title': 'New Dataset',\n",
       " 'Organism': 'New Organism',\n",
       " 'Strain': 'Unknown Strain',\n",
       " 'Number of Samples': 647,\n",
       " 'Number of Unique Conditions': 231,\n",
       " 'Number of Genes': 3906,\n",
       " 'Number of iModulons': 80}"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ica_data.dataset_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Title': 'M. tuberculosis Modulome',\n",
       " 'Organism': '<i>M. tuberculosis</i>',\n",
       " 'Strain': 'H37Rv',\n",
       " 'Number of Samples': 647,\n",
       " 'Number of Unique Conditions': 231,\n",
       " 'Number of Genes': 3906,\n",
       " 'Number of iModulons': 80,\n",
       " 'Useful Additional Link': '<a href=\"https://github.com/Reosu/modulome_mtb\">Github</a>'}"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# FILL IN THESE DETAILS, TOO\n",
    "\n",
    "# This title can combine some information from the splash table\n",
    "ica_data.dataset_table['Title'] = 'M. tuberculosis Modulome'\n",
    "ica_data.dataset_table['Organism'] = '<i>M. tuberculosis</i>'\n",
    "ica_data.dataset_table['Strain'] = 'H37Rv'\n",
    "\n",
    "# additional rows if desired\n",
    "ica_data.dataset_table['Useful Additional Link'] = '<a href=\"https://github.com/Reosu/modulome_mtb\">Github</a>'\n",
    "\n",
    "# check output\n",
    "ica_data.dataset_table"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 3: Add Links (Optional)\n",
    "\n",
    "If your organism has a database like EcoCyc, SubtiWiki, or AureoWiki, you may be interested in adding links to it.\n",
    "\n",
    "iModulonDB has two kinds of links:\n",
    "\n",
    "- gene_links: \n",
    "    - Associate gene **loci** from the **index of ica_data.M** to websites\n",
    "    - On the [gene pages](https://imodulondb.org/gene.html?organism=e_coli&dataset=precise1&gene_id=b0002), these links have the name of the database that you are linking to. The name of the database needs to be saved as a string in ica_data.link_database.\n",
    "    - For programming reasons, the default value of gene_links is all genes as keys, each pointing to the value np.nan.\n",
    "    \n",
    "- tf_links:\n",
    "    - Associate **TF names** from **ica_data.TRN.regulator and ica_data.imodulon_table.regulator** to websites\n",
    "    - On the [iModulon pages](https://imodulondb.org/iModulon.html?organism=e_coli&dataset=precise1&k=43), the name of the regulator becomes a link. You never have to state the name of this database, and it doesn't have to be the same one used for gene_links (i.e. ica_data.link_database does NOT have to be the one you use).\n",
    "    - For programming reasons, the default value is an empty dictionary.\n",
    "    \n",
    "Note that gene_links uses gene loci since not every gene has a readable name, but tf_links uses readable names since not all regulators have associated gene loci. This could create some confusion between the two types of links; for example, the gene b2741 in E. coli encodes the TF rpoS:\n",
    "\n",
    "- ica_data.gene_links\\['b2741'\\] = 'https://ecocyc.org/gene?orgid=ECOLI&id=EG10510 '\n",
    "- ica_data.tf_links\\['rpoS'\\] = 'http://regulondb.ccg.unam.mx/sigmulon?term=ECK125110294&organism=ECK12&format=jsp&type=sigmulon '\n",
    "\n",
    "In this example, ica_data.link_database = 'EcoCyc' since that is where the gene links point to.\n",
    "\n",
    "You are free to skip this step, or you can use any of the following strategies to generate your links:\n",
    "\n",
    "#### Strategy 1: Look at the URLs in your database.\n",
    "\n",
    "Search for genes in your database. You might notice that it always uses the same URL, but enters your search term in part of that URL. You could then take the pieces of that URL and use those to generate all needed URLs. Below is an example from psuedomonas aeruginosa:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "prefix = 'https://mycobrowser.epfl.ch/genes/'\n",
    "suffix = ''\n",
    "\n",
    "temp_gene_links = dict()\n",
    "for gene in ica_data.M.index:\n",
    "    temp_gene_links[gene] = prefix + gene + suffix\n",
    "    \n",
    "temp_tf_links = dict()\n",
    "for gene in ica_data.trn.regulator.unique():\n",
    "    temp_tf_links[gene] = prefix + gene + suffix\n",
    "    \n",
    "# Then, test a few of the links and uncomment these lines\n",
    "\n",
    "ica_data.gene_links = temp_gene_links\n",
    "ica_data.tf_links = temp_tf_links"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You may need to add a suffix in addition to a prefix. You may also have to deal with databases that use different loci abbreviations than the ones you have - this can be remedied by finding a mapping file between the two naming systems.\n",
    "\n",
    "#### Strategy 2: Make a file\n",
    "\n",
    "If you are more proficient in excel than python, have found a gene <-> link mapping file elsewhere, or already used strategy 1 to generate a file, you can make sure it has only two columns (gene and link) and no headers, then simply read it in as shown below.\n",
    "\n",
    "This may be especially useful if you have no database, but want to keep a file of gene_links or tf_links with specific papers instead of db entries."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ica_data.gene_links = 'gene_links.csv'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Strategy 3: Individual modifications\n",
    "\n",
    "You can also just tell the model the links one at a time. This can be useful if you generate all links with another strategy, but want to overide it in a few cases."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ica_data.gene_links['b2741'] = 'https://ecocyc.org/gene?orgid=ECOLI&id=EG10510'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### If using gene links, don't forget to update ica_data.link_database\n",
    "\n",
    "If you don't, the default value is 'External Database'. If you have no links, nothing will appear so this is irrelevant."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'External Database'"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# ica_data.link_database = 'EcoCyc'\n",
    "\n",
    "ica_data.link_database"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 4: Check for compatibility\n",
    "\n",
    "Though the code cannot check for all edge cases and concerns, you can use the function 'imodulondb_compatibility()' to see what information you are missing, mostly through a check of the columns of your DataFrames. \n",
    "\n",
    "#### 4.a: Modify column names\n",
    "\n",
    "First, let's see what we are missing:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "iModulon Table is missing a name column\n",
      "iModulon Table is missing a Regulator column\n",
      "Note that Regulator is displayed in the dataset page so it can have spelled out 'and' and 'or' operators.If you have a TF column, a Regulator column will begenerated for you (but not vice versa).\n",
      "\n",
      "Gene Table is missing the optional length column\n",
      "\n",
      "Sample Table is missing the optional sample column\n",
      "The 0th column will be used to define the names of samples in the activity bar graph unless you add this column.\n"
     ]
    }
   ],
   "source": [
    "imodulondb_compatibility(ica_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now, let's see if we actually DO have the information that the function believes is missing -- it is likely that some of the names of the columns are simply not what iModulonDB is expecting. If so, we simply rename it to what is expected.\n",
    "\n",
    "If you actually DO NOT have some of the information, that is okay -- after this step, we'll fill in the blanks with default values as needed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TF</th>\n",
       "      <th>Function</th>\n",
       "      <th>Category</th>\n",
       "      <th>n_genes</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>category_num</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SigH</th>\n",
       "      <td>sigH</td>\n",
       "      <td>Sigma Factor involved in heat shock and oxidat...</td>\n",
       "      <td>Virulence/Persistence</td>\n",
       "      <td>45</td>\n",
       "      <td>0.082192</td>\n",
       "      <td>0.800000</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Polyketide Synthase Complex</th>\n",
       "      <td>None</td>\n",
       "      <td>KEGG Map related to the synthesis of type II p...</td>\n",
       "      <td>Fatty Acid/Cholestrol</td>\n",
       "      <td>2</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IniR</th>\n",
       "      <td>iniR</td>\n",
       "      <td>Transcription factor that regulates the expres...</td>\n",
       "      <td>Antibiotic</td>\n",
       "      <td>3</td>\n",
       "      <td>0.100000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PyrR</th>\n",
       "      <td>pyrR</td>\n",
       "      <td>Transcription factor that regulates the expres...</td>\n",
       "      <td>Nucleic Acids</td>\n",
       "      <td>44</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WhiB4</th>\n",
       "      <td>whiB4</td>\n",
       "      <td>Redox-Responsive transcription regulator that ...</td>\n",
       "      <td>Redox</td>\n",
       "      <td>95</td>\n",
       "      <td>0.184834</td>\n",
       "      <td>0.410526</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Unc_9</th>\n",
       "      <td>None</td>\n",
       "      <td>Uncharacterized iModulon 9</td>\n",
       "      <td>Uncharacterized</td>\n",
       "      <td>24</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SG_13</th>\n",
       "      <td>None</td>\n",
       "      <td>Single Gene iModulon 13</td>\n",
       "      <td>Single Gene</td>\n",
       "      <td>5</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>76</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Central Carbon Metabolism</th>\n",
       "      <td>None</td>\n",
       "      <td>iModulon enriched with the Central Carbon Meta...</td>\n",
       "      <td>Central Carbon</td>\n",
       "      <td>28</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>77</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Positive Regulation of Growth</th>\n",
       "      <td>None</td>\n",
       "      <td>iModulon enriched with the Positive Regulation...</td>\n",
       "      <td>Growth</td>\n",
       "      <td>26</td>\n",
       "      <td>0.366667</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>78</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WhiB4/IdeR</th>\n",
       "      <td>whiB4/ideR</td>\n",
       "      <td>iModulon that maps to the union of the Rv3681c...</td>\n",
       "      <td>Redox</td>\n",
       "      <td>16</td>\n",
       "      <td>0.230769</td>\n",
       "      <td>0.375000</td>\n",
       "      <td>79</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>80 rows Ã— 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                       TF  \\\n",
       "SigH                                 sigH   \n",
       "Polyketide Synthase Complex          None   \n",
       "IniR                                 iniR   \n",
       "PyrR                                 pyrR   \n",
       "WhiB4                               whiB4   \n",
       "...                                   ...   \n",
       "Unc_9                                None   \n",
       "SG_13                                None   \n",
       "Central Carbon Metabolism            None   \n",
       "Positive Regulation of Growth        None   \n",
       "WhiB4/IdeR                     whiB4/ideR   \n",
       "\n",
       "                                                                        Function  \\\n",
       "SigH                           Sigma Factor involved in heat shock and oxidat...   \n",
       "Polyketide Synthase Complex    KEGG Map related to the synthesis of type II p...   \n",
       "IniR                           Transcription factor that regulates the expres...   \n",
       "PyrR                           Transcription factor that regulates the expres...   \n",
       "WhiB4                          Redox-Responsive transcription regulator that ...   \n",
       "...                                                                          ...   \n",
       "Unc_9                                                 Uncharacterized iModulon 9   \n",
       "SG_13                                                    Single Gene iModulon 13   \n",
       "Central Carbon Metabolism      iModulon enriched with the Central Carbon Meta...   \n",
       "Positive Regulation of Growth  iModulon enriched with the Positive Regulation...   \n",
       "WhiB4/IdeR                     iModulon that maps to the union of the Rv3681c...   \n",
       "\n",
       "                                            Category  n_genes  precision  \\\n",
       "SigH                           Virulence/Persistence       45   0.082192   \n",
       "Polyketide Synthase Complex    Fatty Acid/Cholestrol        2   0.333333   \n",
       "IniR                                      Antibiotic        3   0.100000   \n",
       "PyrR                                   Nucleic Acids       44   0.000000   \n",
       "WhiB4                                          Redox       95   0.184834   \n",
       "...                                              ...      ...        ...   \n",
       "Unc_9                                Uncharacterized       24   0.000000   \n",
       "SG_13                                    Single Gene        5   0.000000   \n",
       "Central Carbon Metabolism             Central Carbon       28   0.000000   \n",
       "Positive Regulation of Growth                 Growth       26   0.366667   \n",
       "WhiB4/IdeR                                     Redox       16   0.230769   \n",
       "\n",
       "                                 recall  category_num  \n",
       "SigH                           0.800000             0  \n",
       "Polyketide Synthase Complex    1.000000             1  \n",
       "IniR                           1.000000             2  \n",
       "PyrR                           0.000000             3  \n",
       "WhiB4                          0.410526             4  \n",
       "...                                 ...           ...  \n",
       "Unc_9                          0.000000            75  \n",
       "SG_13                          0.000000            76  \n",
       "Central Carbon Metabolism      0.000000            77  \n",
       "Positive Regulation of Growth  0.333333            78  \n",
       "WhiB4/IdeR                     0.375000            79  \n",
       "\n",
       "[80 rows x 7 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ica_data.imodulon_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# We do have regulators. Rename that column to TFs\n",
    "# We also have n_genes. Rename that column to n_genes\n",
    "rename_dict = {'regulator': 'TF', # this is more important than the 'Regulator' column\n",
    "               'regulon_size': 'n_genes'} \n",
    "ica_data.imodulon_table = ica_data.imodulon_table.rename(rename_dict, axis = 1)\n",
    "\n",
    "# We also have names. They were in the index, but the code wants them in the main dataframe as well.\n",
    "# If you would rather not add this column, the code will do so automatically.\n",
    "ica_data.imodulon_table['name'] = ica_data.imodulon_table.index\n",
    "\n",
    "# We don't have a separate Regulator column. Don't worry about that!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>accession</th>\n",
       "      <th>old_locus_tag</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_product</th>\n",
       "      <th>COG</th>\n",
       "      <th>uniprot</th>\n",
       "      <th>operon</th>\n",
       "      <th>regulator</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Rv0001</th>\n",
       "      <td>dnaA</td>\n",
       "      <td>NC_000962.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1524</td>\n",
       "      <td>+</td>\n",
       "      <td>chromosomal replication initiator protein DnaA</td>\n",
       "      <td>Replication, recombination and repair</td>\n",
       "      <td>P9WNW3</td>\n",
       "      <td>Op1</td>\n",
       "      <td>Rv0023,Rv0047c,Rv0081,Rv0302,Rv0474,tcrA,Rv067...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rv0002</th>\n",
       "      <td>dnaN</td>\n",
       "      <td>NC_000962.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2052</td>\n",
       "      <td>3260</td>\n",
       "      <td>+</td>\n",
       "      <td>DNA polymerase III subunit beta</td>\n",
       "      <td>Replication, recombination and repair</td>\n",
       "      <td>P9WNU1</td>\n",
       "      <td>Op2</td>\n",
       "      <td>Rv0047c,Rv0081,Rv0135c,Rv0324,Rv0767c,Rv0880,t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rv0003</th>\n",
       "      <td>recF</td>\n",
       "      <td>NC_000962.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3280</td>\n",
       "      <td>4437</td>\n",
       "      <td>+</td>\n",
       "      <td>DNA replication/repair protein RecF</td>\n",
       "      <td>Replication, recombination and repair</td>\n",
       "      <td>P9WHI9</td>\n",
       "      <td>Op2</td>\n",
       "      <td>Rv0023,Rv0081,Rv0302,trcR,Rv1353c,Rv1776c,mbcA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rv0004</th>\n",
       "      <td>Rv0004</td>\n",
       "      <td>NC_000962.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4434</td>\n",
       "      <td>4997</td>\n",
       "      <td>+</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>P9WFL1</td>\n",
       "      <td>Op2</td>\n",
       "      <td>Rv0023,Rv0047c,Rv0324,Rv1473A,kstR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rv0005</th>\n",
       "      <td>gyrB</td>\n",
       "      <td>NC_000962.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5240</td>\n",
       "      <td>7267</td>\n",
       "      <td>+</td>\n",
       "      <td>DNA gyrase subunit B</td>\n",
       "      <td>Replication, recombination and repair</td>\n",
       "      <td>P9WG45</td>\n",
       "      <td>Op3</td>\n",
       "      <td>Rv0023,Rv0047c,Rv0081,Rv0324,Rv0678,csoR,Rv104...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rv3920c</th>\n",
       "      <td>Rv3920c</td>\n",
       "      <td>NC_000962.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4408334</td>\n",
       "      <td>4408897</td>\n",
       "      <td>-</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>O53598</td>\n",
       "      <td>Op2448</td>\n",
       "      <td>Rv0047c,Rv0081,Rv0302,csoR,Rv1353c,cmr,Rv1719,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rv3921c</th>\n",
       "      <td>Rv3921c</td>\n",
       "      <td>NC_000962.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4408969</td>\n",
       "      <td>4410069</td>\n",
       "      <td>-</td>\n",
       "      <td>membrane protein insertase YidC</td>\n",
       "      <td>Intracellular trafficking, secretion, and vesi...</td>\n",
       "      <td>P9WIT5</td>\n",
       "      <td>Op2449</td>\n",
       "      <td>Rv0023,Rv0047c,Rv0081,Rv0135c,Rv0302,Rv0324,cs...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rv3922c</th>\n",
       "      <td>Rv3922c</td>\n",
       "      <td>NC_000962.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4410053</td>\n",
       "      <td>4410415</td>\n",
       "      <td>-</td>\n",
       "      <td>membrane protein insertion efficiency factor</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>P9WFL9</td>\n",
       "      <td>Op2449</td>\n",
       "      <td>Rv2034,lsr2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rv3923c</th>\n",
       "      <td>rnpA</td>\n",
       "      <td>NC_000962.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4410412</td>\n",
       "      <td>4410789</td>\n",
       "      <td>-</td>\n",
       "      <td>ribonuclease P protein component</td>\n",
       "      <td>Translation, ribosomal structure and biogenesis</td>\n",
       "      <td>P9WGZ3</td>\n",
       "      <td>Op2449</td>\n",
       "      <td>Rv0081,Rv0678,Rv1049,Rv1255c,Rv2324,lsr2,tcrX,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rv3924c</th>\n",
       "      <td>rpmH</td>\n",
       "      <td>NC_000962.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4410786</td>\n",
       "      <td>4410929</td>\n",
       "      <td>-</td>\n",
       "      <td>50S ribosomal protein L34</td>\n",
       "      <td>Translation, ribosomal structure and biogenesis</td>\n",
       "      <td>P9WH93</td>\n",
       "      <td>Op2449</td>\n",
       "      <td>Rv0081,trcR,Rv1353c,Rv2250c,devR,espR,MG2; Sta...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3906 rows Ã— 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        gene_name    accession  old_locus_tag    start     stop strand  \\\n",
       "Rv0001       dnaA  NC_000962.3            NaN        1     1524      +   \n",
       "Rv0002       dnaN  NC_000962.3            NaN     2052     3260      +   \n",
       "Rv0003       recF  NC_000962.3            NaN     3280     4437      +   \n",
       "Rv0004     Rv0004  NC_000962.3            NaN     4434     4997      +   \n",
       "Rv0005       gyrB  NC_000962.3            NaN     5240     7267      +   \n",
       "...           ...          ...            ...      ...      ...    ...   \n",
       "Rv3920c   Rv3920c  NC_000962.3            NaN  4408334  4408897      -   \n",
       "Rv3921c   Rv3921c  NC_000962.3            NaN  4408969  4410069      -   \n",
       "Rv3922c   Rv3922c  NC_000962.3            NaN  4410053  4410415      -   \n",
       "Rv3923c      rnpA  NC_000962.3            NaN  4410412  4410789      -   \n",
       "Rv3924c      rpmH  NC_000962.3            NaN  4410786  4410929      -   \n",
       "\n",
       "                                           gene_product  \\\n",
       "Rv0001   chromosomal replication initiator protein DnaA   \n",
       "Rv0002                  DNA polymerase III subunit beta   \n",
       "Rv0003              DNA replication/repair protein RecF   \n",
       "Rv0004                             hypothetical protein   \n",
       "Rv0005                             DNA gyrase subunit B   \n",
       "...                                                 ...   \n",
       "Rv3920c                            hypothetical protein   \n",
       "Rv3921c                 membrane protein insertase YidC   \n",
       "Rv3922c    membrane protein insertion efficiency factor   \n",
       "Rv3923c                ribonuclease P protein component   \n",
       "Rv3924c                       50S ribosomal protein L34   \n",
       "\n",
       "                                                       COG uniprot  operon  \\\n",
       "Rv0001               Replication, recombination and repair  P9WNW3     Op1   \n",
       "Rv0002               Replication, recombination and repair  P9WNU1     Op2   \n",
       "Rv0003               Replication, recombination and repair  P9WHI9     Op2   \n",
       "Rv0004                                    Function unknown  P9WFL1     Op2   \n",
       "Rv0005               Replication, recombination and repair  P9WG45     Op3   \n",
       "...                                                    ...     ...     ...   \n",
       "Rv3920c                                   Function unknown  O53598  Op2448   \n",
       "Rv3921c  Intracellular trafficking, secretion, and vesi...  P9WIT5  Op2449   \n",
       "Rv3922c                                   Function unknown  P9WFL9  Op2449   \n",
       "Rv3923c    Translation, ribosomal structure and biogenesis  P9WGZ3  Op2449   \n",
       "Rv3924c    Translation, ribosomal structure and biogenesis  P9WH93  Op2449   \n",
       "\n",
       "                                                 regulator  \n",
       "Rv0001   Rv0023,Rv0047c,Rv0081,Rv0302,Rv0474,tcrA,Rv067...  \n",
       "Rv0002   Rv0047c,Rv0081,Rv0135c,Rv0324,Rv0767c,Rv0880,t...  \n",
       "Rv0003   Rv0023,Rv0081,Rv0302,trcR,Rv1353c,Rv1776c,mbcA...  \n",
       "Rv0004                  Rv0023,Rv0047c,Rv0324,Rv1473A,kstR  \n",
       "Rv0005   Rv0023,Rv0047c,Rv0081,Rv0324,Rv0678,csoR,Rv104...  \n",
       "...                                                    ...  \n",
       "Rv3920c  Rv0047c,Rv0081,Rv0302,csoR,Rv1353c,cmr,Rv1719,...  \n",
       "Rv3921c  Rv0023,Rv0047c,Rv0081,Rv0135c,Rv0302,Rv0324,cs...  \n",
       "Rv3922c                                        Rv2034,lsr2  \n",
       "Rv3923c  Rv0081,Rv0678,Rv1049,Rv1255c,Rv2324,lsr2,tcrX,...  \n",
       "Rv3924c  Rv0081,trcR,Rv1353c,Rv2250c,devR,espR,MG2; Sta...  \n",
       "\n",
       "[3906 rows x 11 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ica_data.gene_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# we have the \"missing\" cog column:\n",
    "rename_dict = {'cog': 'COG'}\n",
    "ica_data.gene_table = ica_data.gene_table.rename(rename_dict, axis = 1)\n",
    "\n",
    "# we don't have the gene_product column. This is a nice-to-have, but you can ignore it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ReleaseDate</th>\n",
       "      <th>TaxID</th>\n",
       "      <th>ScientificName</th>\n",
       "      <th>species_tax_id</th>\n",
       "      <th>class_tax_id</th>\n",
       "      <th>strain_tax_id</th>\n",
       "      <th>species</th>\n",
       "      <th>LibraryLayout</th>\n",
       "      <th>Platform</th>\n",
       "      <th>Model</th>\n",
       "      <th>...</th>\n",
       "      <th>condition</th>\n",
       "      <th>Biological Replicates</th>\n",
       "      <th>reference_condition</th>\n",
       "      <th>passed_fastqc</th>\n",
       "      <th>passed_reads_mapped_to_CDS</th>\n",
       "      <th>passed_global_correlation</th>\n",
       "      <th>full_name</th>\n",
       "      <th>passed_similar_replicates</th>\n",
       "      <th>passed_number_replicates</th>\n",
       "      <th>DOI</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRX3995807</th>\n",
       "      <td>2018-04-29 10:56:12</td>\n",
       "      <td>83332</td>\n",
       "      <td>Mycobacterium tuberculosis H37Rv</td>\n",
       "      <td>1773</td>\n",
       "      <td>1760</td>\n",
       "      <td>83332</td>\n",
       "      <td>Mycobacterium tuberculosis</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>...</td>\n",
       "      <td>AX_10_MIC</td>\n",
       "      <td>2</td>\n",
       "      <td>wt</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>AX:AX_10_MIC</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>https://dx.doi.org/10.1128%2FmBio.01276-18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX3995806</th>\n",
       "      <td>2018-04-29 10:56:12</td>\n",
       "      <td>83332</td>\n",
       "      <td>Mycobacterium tuberculosis H37Rv</td>\n",
       "      <td>1773</td>\n",
       "      <td>1760</td>\n",
       "      <td>83332</td>\n",
       "      <td>Mycobacterium tuberculosis</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>...</td>\n",
       "      <td>AX_10_MIC</td>\n",
       "      <td>2</td>\n",
       "      <td>wt</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>AX:AX_10_MIC</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>https://dx.doi.org/10.1128%2FmBio.01276-18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX3995809</th>\n",
       "      <td>2018-04-29 10:56:12</td>\n",
       "      <td>83332</td>\n",
       "      <td>Mycobacterium tuberculosis H37Rv</td>\n",
       "      <td>1773</td>\n",
       "      <td>1760</td>\n",
       "      <td>83332</td>\n",
       "      <td>Mycobacterium tuberculosis</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>...</td>\n",
       "      <td>AX_30_MIC</td>\n",
       "      <td>2</td>\n",
       "      <td>wt</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>AX:AX_30_MIC</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>https://dx.doi.org/10.1128%2FmBio.01276-18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX3995808</th>\n",
       "      <td>2018-04-29 10:56:12</td>\n",
       "      <td>83332</td>\n",
       "      <td>Mycobacterium tuberculosis H37Rv</td>\n",
       "      <td>1773</td>\n",
       "      <td>1760</td>\n",
       "      <td>83332</td>\n",
       "      <td>Mycobacterium tuberculosis</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>...</td>\n",
       "      <td>AX_30_MIC</td>\n",
       "      <td>2</td>\n",
       "      <td>wt</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>AX:AX_30_MIC</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>https://dx.doi.org/10.1128%2FmBio.01276-18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX4619048</th>\n",
       "      <td>2019-10-31 17:06:04</td>\n",
       "      <td>83332</td>\n",
       "      <td>Mycobacterium tuberculosis H37Rv</td>\n",
       "      <td>1773</td>\n",
       "      <td>1760</td>\n",
       "      <td>83332</td>\n",
       "      <td>Mycobacterium tuberculosis</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>...</td>\n",
       "      <td>QZ_10_MIC</td>\n",
       "      <td>2</td>\n",
       "      <td>wt</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>AX:QZ_10_MIC</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>https://dx.doi.org/10.1128%2FmBio.01276-18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX4498338</th>\n",
       "      <td>2018-09-25 17:56:05</td>\n",
       "      <td>1773</td>\n",
       "      <td>Mycobacterium tuberculosis</td>\n",
       "      <td>1773</td>\n",
       "      <td>1760</td>\n",
       "      <td>1773</td>\n",
       "      <td>Mycobacterium tuberculosis</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>NextSeq 500</td>\n",
       "      <td>...</td>\n",
       "      <td>1_days_176</td>\n",
       "      <td>3</td>\n",
       "      <td>0_days_control</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>vera:1_days_176</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>https://doi.org/10.1128/mbio.02627-19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX4498329</th>\n",
       "      <td>2018-09-25 17:56:05</td>\n",
       "      <td>1773</td>\n",
       "      <td>Mycobacterium tuberculosis</td>\n",
       "      <td>1773</td>\n",
       "      <td>1760</td>\n",
       "      <td>1773</td>\n",
       "      <td>Mycobacterium tuberculosis</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>NextSeq 500</td>\n",
       "      <td>...</td>\n",
       "      <td>1_days_control</td>\n",
       "      <td>4</td>\n",
       "      <td>0_days_control</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>vera:1_days_control</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>https://doi.org/10.1128/mbio.02627-19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX4498325</th>\n",
       "      <td>2018-09-25 17:56:05</td>\n",
       "      <td>1773</td>\n",
       "      <td>Mycobacterium tuberculosis</td>\n",
       "      <td>1773</td>\n",
       "      <td>1760</td>\n",
       "      <td>1773</td>\n",
       "      <td>Mycobacterium tuberculosis</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>NextSeq 500</td>\n",
       "      <td>...</td>\n",
       "      <td>1_days_control</td>\n",
       "      <td>4</td>\n",
       "      <td>0_days_control</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>vera:1_days_control</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>https://doi.org/10.1128/mbio.02627-19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX4498327</th>\n",
       "      <td>2018-09-25 17:56:05</td>\n",
       "      <td>1773</td>\n",
       "      <td>Mycobacterium tuberculosis</td>\n",
       "      <td>1773</td>\n",
       "      <td>1760</td>\n",
       "      <td>1773</td>\n",
       "      <td>Mycobacterium tuberculosis</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>NextSeq 500</td>\n",
       "      <td>...</td>\n",
       "      <td>1_days_control</td>\n",
       "      <td>4</td>\n",
       "      <td>0_days_control</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>vera:1_days_control</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>https://doi.org/10.1128/mbio.02627-19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX4498331</th>\n",
       "      <td>2018-09-25 17:56:05</td>\n",
       "      <td>1773</td>\n",
       "      <td>Mycobacterium tuberculosis</td>\n",
       "      <td>1773</td>\n",
       "      <td>1760</td>\n",
       "      <td>1773</td>\n",
       "      <td>Mycobacterium tuberculosis</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>NextSeq 500</td>\n",
       "      <td>...</td>\n",
       "      <td>1_days_control</td>\n",
       "      <td>4</td>\n",
       "      <td>0_days_control</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>vera:1_days_control</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>https://doi.org/10.1128/mbio.02627-19</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>647 rows Ã— 44 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    ReleaseDate  TaxID                    ScientificName  \\\n",
       "SRX3995807  2018-04-29 10:56:12  83332  Mycobacterium tuberculosis H37Rv   \n",
       "SRX3995806  2018-04-29 10:56:12  83332  Mycobacterium tuberculosis H37Rv   \n",
       "SRX3995809  2018-04-29 10:56:12  83332  Mycobacterium tuberculosis H37Rv   \n",
       "SRX3995808  2018-04-29 10:56:12  83332  Mycobacterium tuberculosis H37Rv   \n",
       "SRX4619048  2019-10-31 17:06:04  83332  Mycobacterium tuberculosis H37Rv   \n",
       "...                         ...    ...                               ...   \n",
       "SRX4498338  2018-09-25 17:56:05   1773        Mycobacterium tuberculosis   \n",
       "SRX4498329  2018-09-25 17:56:05   1773        Mycobacterium tuberculosis   \n",
       "SRX4498325  2018-09-25 17:56:05   1773        Mycobacterium tuberculosis   \n",
       "SRX4498327  2018-09-25 17:56:05   1773        Mycobacterium tuberculosis   \n",
       "SRX4498331  2018-09-25 17:56:05   1773        Mycobacterium tuberculosis   \n",
       "\n",
       "            species_tax_id  class_tax_id  strain_tax_id  \\\n",
       "SRX3995807            1773          1760          83332   \n",
       "SRX3995806            1773          1760          83332   \n",
       "SRX3995809            1773          1760          83332   \n",
       "SRX3995808            1773          1760          83332   \n",
       "SRX4619048            1773          1760          83332   \n",
       "...                    ...           ...            ...   \n",
       "SRX4498338            1773          1760           1773   \n",
       "SRX4498329            1773          1760           1773   \n",
       "SRX4498325            1773          1760           1773   \n",
       "SRX4498327            1773          1760           1773   \n",
       "SRX4498331            1773          1760           1773   \n",
       "\n",
       "                               species LibraryLayout  Platform  \\\n",
       "SRX3995807  Mycobacterium tuberculosis        SINGLE  ILLUMINA   \n",
       "SRX3995806  Mycobacterium tuberculosis        SINGLE  ILLUMINA   \n",
       "SRX3995809  Mycobacterium tuberculosis        SINGLE  ILLUMINA   \n",
       "SRX3995808  Mycobacterium tuberculosis        SINGLE  ILLUMINA   \n",
       "SRX4619048  Mycobacterium tuberculosis        SINGLE  ILLUMINA   \n",
       "...                                ...           ...       ...   \n",
       "SRX4498338  Mycobacterium tuberculosis        SINGLE  ILLUMINA   \n",
       "SRX4498329  Mycobacterium tuberculosis        SINGLE  ILLUMINA   \n",
       "SRX4498325  Mycobacterium tuberculosis        SINGLE  ILLUMINA   \n",
       "SRX4498327  Mycobacterium tuberculosis        SINGLE  ILLUMINA   \n",
       "SRX4498331  Mycobacterium tuberculosis        SINGLE  ILLUMINA   \n",
       "\n",
       "                          Model  ...       condition Biological Replicates  \\\n",
       "SRX3995807  Illumina HiSeq 2500  ...       AX_10_MIC                     2   \n",
       "SRX3995806  Illumina HiSeq 2500  ...       AX_10_MIC                     2   \n",
       "SRX3995809  Illumina HiSeq 2500  ...       AX_30_MIC                     2   \n",
       "SRX3995808  Illumina HiSeq 2500  ...       AX_30_MIC                     2   \n",
       "SRX4619048  Illumina HiSeq 2500  ...       QZ_10_MIC                     2   \n",
       "...                         ...  ...             ...                   ...   \n",
       "SRX4498338          NextSeq 500  ...      1_days_176                     3   \n",
       "SRX4498329          NextSeq 500  ...  1_days_control                     4   \n",
       "SRX4498325          NextSeq 500  ...  1_days_control                     4   \n",
       "SRX4498327          NextSeq 500  ...  1_days_control                     4   \n",
       "SRX4498331          NextSeq 500  ...  1_days_control                     4   \n",
       "\n",
       "           reference_condition  passed_fastqc passed_reads_mapped_to_CDS  \\\n",
       "SRX3995807                  wt           True                       True   \n",
       "SRX3995806                  wt           True                       True   \n",
       "SRX3995809                  wt           True                       True   \n",
       "SRX3995808                  wt           True                       True   \n",
       "SRX4619048                  wt           True                       True   \n",
       "...                        ...            ...                        ...   \n",
       "SRX4498338      0_days_control           True                       True   \n",
       "SRX4498329      0_days_control           True                       True   \n",
       "SRX4498325      0_days_control           True                       True   \n",
       "SRX4498327      0_days_control           True                       True   \n",
       "SRX4498331      0_days_control           True                       True   \n",
       "\n",
       "           passed_global_correlation            full_name  \\\n",
       "SRX3995807                      True         AX:AX_10_MIC   \n",
       "SRX3995806                      True         AX:AX_10_MIC   \n",
       "SRX3995809                      True         AX:AX_30_MIC   \n",
       "SRX3995808                      True         AX:AX_30_MIC   \n",
       "SRX4619048                      True         AX:QZ_10_MIC   \n",
       "...                              ...                  ...   \n",
       "SRX4498338                      True      vera:1_days_176   \n",
       "SRX4498329                      True  vera:1_days_control   \n",
       "SRX4498325                      True  vera:1_days_control   \n",
       "SRX4498327                      True  vera:1_days_control   \n",
       "SRX4498331                      True  vera:1_days_control   \n",
       "\n",
       "           passed_similar_replicates passed_number_replicates  \\\n",
       "SRX3995807                      True                     True   \n",
       "SRX3995806                      True                     True   \n",
       "SRX3995809                      True                     True   \n",
       "SRX3995808                      True                     True   \n",
       "SRX4619048                      True                     True   \n",
       "...                              ...                      ...   \n",
       "SRX4498338                      True                     True   \n",
       "SRX4498329                      True                     True   \n",
       "SRX4498325                      True                     True   \n",
       "SRX4498327                      True                     True   \n",
       "SRX4498331                      True                     True   \n",
       "\n",
       "                                                   DOI  \n",
       "SRX3995807  https://dx.doi.org/10.1128%2FmBio.01276-18  \n",
       "SRX3995806  https://dx.doi.org/10.1128%2FmBio.01276-18  \n",
       "SRX3995809  https://dx.doi.org/10.1128%2FmBio.01276-18  \n",
       "SRX3995808  https://dx.doi.org/10.1128%2FmBio.01276-18  \n",
       "SRX4619048  https://dx.doi.org/10.1128%2FmBio.01276-18  \n",
       "...                                                ...  \n",
       "SRX4498338       https://doi.org/10.1128/mbio.02627-19  \n",
       "SRX4498329       https://doi.org/10.1128/mbio.02627-19  \n",
       "SRX4498325       https://doi.org/10.1128/mbio.02627-19  \n",
       "SRX4498327       https://doi.org/10.1128/mbio.02627-19  \n",
       "SRX4498331       https://doi.org/10.1128/mbio.02627-19  \n",
       "\n",
       "[647 rows x 44 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ica_data.sample_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Nothing to change for the sample table -- the 0th column is the correct one to use.\n",
    "# However, you can optionally suppress a warning about this from later on if you name the index:\n",
    "ica_data.sample_table.index.name = 'sample'\n",
    "\n",
    "# Also, there is an error in the existing 'Biological Replicates' column.\n",
    "# This caused a \"cannot set a row with mismatched columns\" error in export (step 5).\n",
    "# The easiest way to fix it is to have the code re-compute this column in step 4b.\n",
    "# to make that happen, delete the existing one now.\n",
    "#ica_data.sample_table = ica_data.sample_table.drop('Biological Replicates', axis = 1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4.b: Fill in the unknowns (Optional)\n",
    "\n",
    "This step is already integrated in step 5 -- feel free to skip it. However, it might be a useful demonstration.\n",
    "\n",
    "Even if you only have the minimum (M, A, X, and sample_table) data, calling the compatibility function with write=True will fill in enough blanks to generate the dataset. In this example, we need the code to fix the things that we ignored during step 4a.\n",
    "\n",
    "In general, you may want to deepcopy the data before doing this. While the default columns shouldn't be an issue, the function requires that all iModulon names be numbers, which might not be desired for you. Under the hood in step 5, we automatically deep copy your data before changing anything. If you want access to the default values and columns, however, you may be interested in doing this without the deepcopy."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "iModulon Table is missing a Regulator column\n",
      "Note that Regulator is displayed in the dataset page so it can have spelled out 'and' and 'or' operators.If you have a TF column, a Regulator column will begenerated for you (but not vice versa).\n",
      "\n",
      "Gene Table is missing the optional length column\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from copy import deepcopy\n",
    "\n",
    "new_data = deepcopy(ica_data)\n",
    "\n",
    "imodulondb_compatibility(new_data, inplace = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's call it again just to see what the output is now that the new elements are written:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Gene Table is missing the optional length column\n",
      "\n"
     ]
    }
   ],
   "source": [
    "imodulondb_compatibility(new_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A very minimal output is good!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5: Export\n",
    "\n",
    "You are now ready to export! This process may take some time (up to around 1 hr for PRECISE2, but significantly shorter depending on the number of genes your organism has).\n",
    "\n",
    "The export function needs a path to export to. This should be path to where you cloned the iModulonDB repository.\n",
    "\n",
    "The function will first deepcopy your files and then call the compatibility function from above with write = True. Then it will generate all files that don't require iteration, then all iModulon files, then all gene files.\n",
    "\n",
    "As it goes, it will print anything you may want to be aware of. This is where issues with the TRN may become apparent."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "iModulon Table is missing a Regulator column\n",
      "Note that Regulator is displayed in the dataset page so it can have spelled out 'and' and 'or' operators.If you have a TF column, a Regulator column will begenerated for you (but not vice versa).\n",
      "\n",
      "Gene Table is missing the optional length column\n",
      "\n",
      "Writing main site files...\n",
      "Two progress bars will appear below. The second will take significantly longer than the first.\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "c427e4c951a443f08f7243b2d460eb5b",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(FloatProgress(value=0.0, max=80.0), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TF not in TRN: lsyG\n",
      "To remedy this, add rows to the TRN for each gene associated with this TF\n",
      "TF has no associated expression profile: lsyG\n",
      "If lsyG is not a gene, this behavior is expected.\n",
      "If it is a gene, use consistent naming between the TRN and gene_table.\n",
      "\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "4737957336d34884bff9e50b0cae7485",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(FloatProgress(value=0.0, max=3906.0), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "imodulondb_export(ica_data, '../iModulonDB/')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 6: Update TRN, other gene info if desired\n",
    "\n",
    "In the output above, you will notice snippets such as this one:\n",
    "\n",
    "> TF has no associated expression profile: cecR\n",
    "\n",
    ">If cecR is not a gene, this behavior is expected.\n",
    "\n",
    ">If it is a gene, use consistent naming between the TRN and gene_table.\n",
    "\n",
    "If desired, you could look up each of these to find alternative names that may be used, then rename either the TRN (all instances of the name in the regulator column) or the gene_table (the name column in the row with the appropriate locus) so that they all match. This will not need to be done for every case; for example, 'ile-tRNA' is a leucine-bound tRNA molecule, which does not have an associated gene.\n",
    "\n",
    "After doing this, you would need to rerun the export function (step 5)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# according to EcoCyc, the b-number for cecR is b0796 and its alternative name is 'ybiH'\n",
    "\n",
    "#ica_data.gene_table.loc['b0796']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "# To remedy this matchup, I will replace 'ybiH' with 'cecR' in the gene_table.\n",
    "\n",
    "#ica_data.gene_table.loc['b0796', 'gene_name'] = 'cecR'\n",
    "\n",
    "# repeat this process for all unmatched genes.\n",
    "\n",
    "# re-running step 5 would then correct this error."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 7: Paste HTML\n",
    "\n",
    "This step only needs to occur once! Time to add the link to the splash page.\n",
    "\n",
    "1. Navigate to your iModulonDB/organisms/M_favoritespeciesus/modulome folder.\n",
    "2. Open the file 'html_for_splash.html' in a text editor\n",
    "3. Copy the entire contents to your clipboard\n",
    "4. Navigate back to the main repository folder\n",
    "5. Open 'index.html' in a text editor\n",
    "6. Scroll down to the comment near line 250 about \"NEW DATASETS\"\n",
    "7. Paste in the appropriate location. You may want to adjust the indentations.\n",
    "8. Save the file\n",
    "9. Enjoy your new website!\n",
    "\n",
    "Traveling to localhost:8000 in your browser should show you the site with a shiny new button for your dataset. Click it to see your new iModulons."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Moving Forward\n",
    "\n",
    "Any time you make a change that you want reflected on your site, simply repeat step 5! If changes don't immediately appear on the browser, try a force refresh (Ctrl+Shift+R in Google Chrome)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Don't forget to save your file so that step 5 is all that is required in the future!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "save_to_json(ica_data, 'mtb_data.json')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.2"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
